uniffi_bindgen/src/interface/types.rs

```diff
@@ -128,243 +132,349 @@ impl From<&Type> for FFIType {
-            weedle::Definition::Typedef(d) => d.find_type_definitions(ci),
+            weedle::Definition::Interface(d) => d.add_type_definitions_to(types),
+            weedle::Definition::Dictionary(d) => d.add_type_definitions_to(types),
+            weedle::Definition::Enum(d) => d.add_type_definitions_to(types),
+            weedle::Definition::Typedef(d) => d.add_type_definitions_to(types),
             _ => Ok(()),
         }
     }
 }
 
 impl TypeFinder for weedle::InterfaceDefinition<'_> {
-    fn find_type_definitions(&self, ci: &mut ComponentInterface) -> Result<()> {
+    fn add_type_definitions_to(&self, types: &mut TypeUniverse) -> Result<()> {
         let name = self.identifier.0.to_string();
-        if resolve_builtin_type(&self.identifier).is_some() {
-            bail!("please don't shadow builtin types (interface {})", name);
-        }
-        ci.add_type_definition(self.identifier.0, Type::Object(name))
+        types.add_type_definition(self.identifier.0, Type::Object(name))
     }
 }
 
 impl TypeFinder for weedle::DictionaryDefinition<'_> {
-    fn find_type_definitions(&self, ci: &mut ComponentInterface) -> Result<()> {
+    fn add_type_definitions_to(&self, types: &mut TypeUniverse) -> Result<()> {
         let name = self.identifier.0.to_string();
-        if resolve_builtin_type(&self.identifier).is_some() {
-            bail!("please don't shadow builtin types (dictionary {})", name);
-        }
-        ci.add_type_definition(self.identifier.0, Type::Record(name))
+        types.add_type_definition(self.identifier.0, Type::Record(name))
     }
 }
 
 impl TypeFinder for weedle::EnumDefinition<'_> {
-    fn find_type_definitions(&self, ci: &mut ComponentInterface) -> Result<()> {
+    fn add_type_definitions_to(&self, types: &mut TypeUniverse) -> Result<()> {
+        let name = self.identifier.0.to_string();
+        // Our error types are defined using an `enum` with a special attribute.
         if let Some(attrs) = &self.attributes {
             let attrs = Attributes::try_from(attrs)?;
             if attrs.contains_error_attr() {
-                let name = self.identifier.0.to_string();
-                return ci.add_type_definition(self.identifier.0, Type::Error(name));
+                return types.add_type_definition(self.identifier.0, Type::Error(name));
             }
         }
-        let name = self.identifier.0.to_string();
-        if resolve_builtin_type(&self.identifier).is_some() {
-            bail!("please don't shadow builtin types (enum {})", name);
-        }
-        ci.add_type_definition(self.identifier.0, Type::Enum(name))
+        types.add_type_definition(self.identifier.0, Type::Enum(name))
     }
 }
 
 impl TypeFinder for weedle::TypedefDefinition<'_> {
-    fn find_type_definitions(&self, ci: &mut ComponentInterface) -> Result<()> {
+    fn add_type_definitions_to(&self, types: &mut TypeUniverse) -> Result<()> {
```

Comment:
Most of the changes in the `TypeFinder` and `TypeResolve` trait impls are simple renamings, I don't think I've made any substantial changes to the logic. (Apart from moving the "dont override builtin names" check into `TypeUniverse.add_type_definition` so that it's done in a single consistent spot).